Deserialize by using __new__ instead of constructor if possible

oldium opened the pull request:
Exception to IOError is because this built-in exception is assigning internal `myerrno` attribute in `__init__` constructor and cannot be changed from Python code.

Fixes: #65

ionelmc commented on src/tblib/pickling_support.py:
```diff
@@ -49,8 +62,12 @@ def pickle_exception(obj):
+
     return (
-        unpickle_exception,
+        unpickle_builtin_exception if use_builtin else unpickler_exception,
```
Is this a typo? There is no `unpickler_exception`. Either way, you need to add a bunch of tests cause this should cause failures.

oldium replied:
Yes, typo. Sorry. Fixed.

Regarding tests - I do not have much time currently, so when time allows, I will add some tests for all situations mentioned in the `use_builtin` check (i.e. custom `__reduce__`, `__reduce_ex__`, `IOError`, “normal” exception).

oldium commented:
Applied changes suggested by the check job.

oldium commented:
Tests added

oldium commented:
Looks like PyPy does not like test for built-in reducer. I will check it.

oldium commented:
I created a bug report for PyPy https://foss.heptapod.net/pypy/pypy/-/issues/4027 regarding the builtin method check. For now, we could switch to `inspect.isbuiltin()` (it looks it is not only a [BuiltinFunctionType check](https://foss.heptapod.net/pypy/pypy/-/blob/branch/default/lib-python/2.7/inspect.py#L237) in PyPy) and skip the test on PyPy until it is fixed?

oldium commented:
I did what I proposed - there are two more commits, one switching to `inspect.isbuiltin()` (i.e. no functional change) and second one disabling issue #64 test. This means that under PyPy the behaviour of `tblib` is unchanged, the bug is still there. I hope PyPy team fixes/addresses the issue soon.

oldium commented:
In the [PyPy bug report](https://foss.heptapod.net/pypy/pypy/-/issues/4027) one suggestion was to use `obj.__class__.__reduce__ is BaseException.__reduce__` and it works both in CPython and PyPy. See the latest commit, which fixes the problem of #64 also for PyPy. Now the tests should be all green 😁

oldium commented:
I am done with this PR.

oldium commented:
Squashed related commits, updated reasoning in description and updated naming in code (no functional change, just added one more test). Also added two fixes for the `tox` build job, feel free to cherry-pick it separately 😊

oldium commented:
Rebased to latest master

oldium commented:
Rebased on the latest master and as a bonus fixed build with Python 3.8, 3.9 and 3.10 on macos.

Enjoy 😁

stuaxo commented:
@oldium not sure if you noticed, but PyPy fixed the isbultin bug you reported.

oldium commented:
@stuaxo Are you sure it is fixed, or you found that the bug is closed?

stuaxo commented:
@oldium ooh, good point, looking at the ticket again I am not sure - apologies for the noise.

oldium commented:
Rebased on top of #79

oldium commented:
FYI my pull request contains all commits from #79, they were necessary to make tests pass.

ionelmc commented:
Rebased in https://github.com/ionelmc/python-tblib/pull/82 but I added a couple more tests and I didn't like how it worked. I have reworked it a bit to handle more situations with broken exception subclasses. Let me know what you think or how it works for you.
